File: test/core/entitlements/entitlement_resolver_test.dart
import 'package:akshara_erp/core/entitlements/entitlement_models.dart';
import 'package:akshara_erp/core/entitlements/entitlement_resolver.dart';
import 'package:akshara_erp/core/school_config/school_capability_registry.dart';
import 'package:akshara_erp/core/school_config/school_configuration_models.dart';
import 'package:flutter_test/flutter_test.dart';

/// Plan entitlement sets mirroring the seed (migration 20260717000000).
const _core = {
  EntitlementSlugs.admissions,
  EntitlementSlugs.finance,
  EntitlementSlugs.sis,
  EntitlementSlugs.management,
  EntitlementSlugs.attendance,
  EntitlementSlugs.exams,
};
const _ops = {
  EntitlementSlugs.transport,
  EntitlementSlugs.hostel,
  EntitlementSlugs.library,
  EntitlementSlugs.inventory,
  EntitlementSlugs.alumni,
  EntitlementSlugs.hrPayroll,
  EntitlementSlugs.multiBranch,
};
final trialPlan = {..._core};
final standardPlan = {..._core};
final professionalPlan = {..._core, ..._ops, EntitlementSlugs.parentInsights};
final enterprisePlan = {
  ...professionalPlan,
  EntitlementSlugs.trustOrg,
  EntitlementSlugs.aiPredictions,
};

/// A school that has switched every module on (the widest local config).
const allOnLocal = SchoolCapabilities(
  transport: true,
  hostel: true,
  library: true,
  inventory: true,
  alumni: true,
  hrPayroll: true,
  multiBranch: true,
  trustOrganization: true,
);

void main() {
  group('plan ceiling × school config = effective capabilities', () {
    test('Trial: all optional modules plan-locked even with everything on', () {
      final caps = EntitlementResolver.resolveEffective(
        local: allOnLocal,
        entitlements: trialPlan,
      );
      expect(caps.transport, isFalse);
      expect(caps.hostel, isFalse);
      expect(caps.library, isFalse);
      expect(caps.inventory, isFalse);
      expect(caps.alumni, isFalse);
      expect(caps.hrPayroll, isFalse);
      expect(caps.multiBranch, isFalse);
      expect(caps.trustOrganization, isFalse);
      // Registry: only core modules remain.
      final ids = SchoolCapabilityRegistry.enabledModuleIds(caps);
      expect(ids, containsAll(['admissions', 'finance', 'sis', 'management']));
      expect(ids, isNot(contains('transport')));
      expect(ids, isNot(contains('hr')));
    });

    test('Standard: core only — ops still plan-locked', () {
      final caps = EntitlementResolver.resolveEffective(
        local: allOnLocal,
        entitlements: standardPlan,
      );
      expect(caps.transport, isFalse);
      expect(caps.multiBranch, isFalse);
      expect(caps.trustOrganization, isFalse);
    });

    test('Professional: ops + multiBranch on, trust still locked', () {
      final caps = EntitlementResolver.resolveEffective(
        local: allOnLocal,
        entitlements: professionalPlan,
      );
      expect(caps.transport, isTrue);
      expect(caps.hostel, isTrue);
      expect(caps.library, isTrue);
      expect(caps.inventory, isTrue);
      expect(caps.alumni, isTrue);
      expect(caps.hrPayroll, isTrue);
      expect(caps.multiBranch, isTrue);
      expect(caps.trustOrganization, isFalse);
      final ids = SchoolCapabilityRegistry.enabledModuleIds(caps);
      expect(ids, containsAll(['transport', 'hostel', 'hr', 'branch']));
      expect(ids, isNot(contains('director')));
    });

    test('Enterprise: every capability available (AI predictions included)', () {
      final caps = EntitlementResolver.resolveEffective(
        local: allOnLocal,
        entitlements: enterprisePlan,
      );
      expect(caps.transport, isTrue);
      expect(caps.multiBranch, isTrue);
      expect(caps.trustOrganization, isTrue);
      expect(enterprisePlan.contains(EntitlementSlugs.aiPredictions), isTrue);
      final ids = SchoolCapabilityRegistry.enabledModuleIds(caps);
      expect(ids, containsAll(['director', 'control_center', 'trust_intelligence']));
    });

    test('School narrows within plan: Professional but transport off locally', () {
      final caps = EntitlementResolver.resolveEffective(
        local: allOnLocal.copyWith(transport: false),
        entitlements: professionalPlan,
      );
      expect(caps.transport, isFalse); // school chose to disable
      expect(caps.hostel, isTrue); // others stay
    });

    test('School can never exceed plan: Standard + transport on locally', () {
      final caps = EntitlementResolver.resolveEffective(
        local: allOnLocal,
        entitlements: standardPlan,
      );
      expect(caps.transport, isFalse);
    });
  });

  group('Trial fallback', () {
    test('trialFallback grants no optional modules', () {
      final sub = ResolvedSubscription.trialFallback();
      expect(sub.planSlug, 'trial');
      expect(sub.status, SubscriptionStatus.trial);
      expect(sub.fallbackApplied, isTrue);
      expect(sub.entitlements, isEmpty);

      final caps = EntitlementResolver.resolveEffective(
        local: allOnLocal,
        entitlements: sub.entitlementSet,
      );
      expect(caps.transport, isFalse);
      expect(caps.trustOrganization, isFalse);
    });
  });

  group('unrestricted ceiling (entitlement layer disabled)', () {
    test('intersect with unrestricted ceiling preserves local config', () {
      final local = allOnLocal.copyWith(multiBranch: false, hostel: false);
      final caps = EntitlementResolver.intersect(
        local,
        EntitlementResolver.unrestrictedCeiling,
      );
      expect(caps.transport, isTrue);
      expect(caps.hostel, isFalse); // local off stays off
      expect(caps.multiBranch, isFalse);
      expect(caps.trustOrganization, isTrue);
    });
  });
}

File: supabase/functions/_shared/entitlements/entitlement_resolver_test.ts
import { assertEquals } from "https://deno.land/std@0.224.0/assert/mod.ts";
import {
  CAPABILITY_ENTITLEMENTS,
  planAllowedEntitlements,
  planAllows,
  resolveEffectiveCapabilities,
  schoolConfigEnabled,
} from "./entitlement_resolver.ts";

// Seed entitlement sets mirroring plan_entitlements (migration 20260717000000).
const CORE = [
  "module.admissions",
  "module.finance",
  "module.sis",
  "module.management",
  "module.attendance",
  "module.exams",
];
const OPS = [
  "module.transport",
  "module.hostel",
  "module.library",
  "module.inventory",
  "module.alumni",
  "module.hr_payroll",
  "module.multi_branch",
];
const TRIAL = CORE;
const STANDARD = CORE;
const PROFESSIONAL = [...CORE, ...OPS, "feature.parent_insights"];
const ENTERPRISE = [
  ...PROFESSIONAL,
  "module.trust_org",
  "feature.ai_predictions",
];

// All 8 flags ON, mirroring a school that toggled everything on.
const ALL_ON: Record<string, boolean> = {
  transport: true,
  hostel: true,
  library: true,
  inventory: true,
  alumni: true,
  hrPayroll: true,
  multiBranch: true,
  trustOrganization: true,
};

// ─── Trial ──────────────────────────────────────────────────────────────────
Deno.test("Trial: core only — every optional capability is plan-locked", () => {
  const caps = resolveEffectiveCapabilities({
    planEntitlements: TRIAL,
    schoolCapabilities: ALL_ON,
  });
  // None of the 8 optional flags are in Trial's plan, so all resolve false even
  // though the school switched them all on.
  for (const flag of Object.keys(CAPABILITY_ENTITLEMENTS)) {
    assertEquals(caps[flag], false, `${flag} must be plan-locked on Trial`);
  }
});

// ─── Standard ─────────────────────────────────────────────────────────────────
Deno.test("Standard: core only — ops modules plan-locked", () => {
  const caps = resolveEffectiveCapabilities({
    planEntitlements: STANDARD,
    schoolCapabilities: ALL_ON,
  });
  assertEquals(caps.transport, false);
  assertEquals(caps.hostel, false);
  assertEquals(caps.multiBranch, false);
  assertEquals(caps.trustOrganization, false);
  // Core modules are never in the capability map, so they are not gated here.
  assertEquals(planAllows("module.finance", STANDARD), true);
});

// ─── Professional ────────────────────────────────────────────────────────────
Deno.test("Professional: ops + multiBranch on, trust still locked", () => {
  const caps = resolveEffectiveCapabilities({
    planEntitlements: PROFESSIONAL,
    schoolCapabilities: ALL_ON,
  });
  assertEquals(caps.transport, true);
  assertEquals(caps.hostel, true);
  assertEquals(caps.library, true);
  assertEquals(caps.inventory, true);
  assertEquals(caps.alumni, true);
  assertEquals(caps.hrPayroll, true);
  assertEquals(caps.multiBranch, true);
  // trust_org is Enterprise-only.
  assertEquals(caps.trustOrganization, false);
  assertEquals(planAllows("feature.parent_insights", PROFESSIONAL), true);
  assertEquals(planAllows("feature.ai_predictions", PROFESSIONAL), false);
});

// ─── Enterprise ──────────────────────────────────────────────────────────────
Deno.test("Enterprise: all 8 capabilities on; AI predictions included", () => {
  const caps = resolveEffectiveCapabilities({
    planEntitlements: ENTERPRISE,
    schoolCapabilities: ALL_ON,
  });
  for (const flag of Object.keys(CAPABILITY_ENTITLEMENTS)) {
    assertEquals(caps[flag], true, `${flag} must be available on Enterprise`);
  }
  // AI predictions is Enterprise-included (not an addon), per locked config.
  assertEquals(planAllows("feature.ai_predictions", ENTERPRISE), true);
});

// ─── School-level disable override (narrow within plan) ──────────────────────
Deno.test("School disable: plan allows transport but school switched it off", () => {
  const caps = resolveEffectiveCapabilities({
    planEntitlements: PROFESSIONAL,
    schoolCapabilities: { ...ALL_ON, transport: false },
  });
  // School narrows within its plan — transport off despite the plan allowing it.
  assertEquals(caps.transport, false);
  // Other Professional capabilities stay on.
  assertEquals(caps.hostel, true);
});

Deno.test("School can never exceed plan: enabling a flag the plan lacks stays false", () => {
  const caps = resolveEffectiveCapabilities({
    planEntitlements: STANDARD,
    schoolCapabilities: { ...ALL_ON, transport: true },
  });
  assertEquals(caps.transport, false);
});

// ─── Missing subscription → Trial fallback ───────────────────────────────────
// The service substitutes Trial's entitlements when no subscription row exists;
// here we prove the resolver yields the locked-down Trial result for that set.
Deno.test("Missing subscription → Trial entitlements lock all optional modules", () => {
  const caps = resolveEffectiveCapabilities({
    planEntitlements: TRIAL, // service supplies Trial's entitlements on no-row
    schoolCapabilities: ALL_ON,
  });
  for (const flag of Object.keys(CAPABILITY_ENTITLEMENTS)) {
    assertEquals(caps[flag], false);
  }
});

// ─── Enterprise per-deal overrides ───────────────────────────────────────────
Deno.test("Override grant adds a slug beyond the plan", () => {
  const allowed = planAllowedEntitlements(PROFESSIONAL, {
    grant: ["feature.ai_predictions"],
  });
  assertEquals(allowed.has("feature.ai_predictions"), true);
});

Deno.test("Override revoke removes a slug the plan grants", () => {
  const allowed = planAllowedEntitlements(ENTERPRISE, {
    revoke: ["module.trust_org"],
  });
  assertEquals(allowed.has("module.trust_org"), false);
});

// ─── Default-capability behaviour (no school_configuration row) ───────────────
Deno.test("No school config row: defaults apply (ops on, multiBranch/trust off)", () => {
  assertEquals(schoolConfigEnabled("transport", null), true);
  assertEquals(schoolConfigEnabled("multiBranch", null), false);
  assertEquals(schoolConfigEnabled("trustOrganization", undefined), false);
});

Deno.test("Professional with no school config: ops resolve on via defaults", () => {
  const caps = resolveEffectiveCapabilities({
    planEntitlements: PROFESSIONAL,
    schoolCapabilities: null,
  });
  assertEquals(caps.transport, true); // plan allows + default on
  assertEquals(caps.multiBranch, false); // plan allows but default off
});
